core/referral/engine.go

```diff
@@ -255,47 +307,64 @@ func (e *Engine) loadReferralSetsFromSnapshot(sets *snapshotpb.ReferralSets) {
 	}
 
 	for _, set := range sets.Sets {
+		setID := types.ReferralSetID(set.Id)
+
 		newSet := &types.ReferralSet{
-			ID:        set.Id,
+			ID:        setID,
 			CreatedAt: time.Unix(0, set.CreatedAt),
 			UpdatedAt: time.Unix(0, set.CreatedAt),
 			Referrer: &types.Membership{
-				PartyID:       types.PartyID(set.Referrer.PartyId),
-				JoinedAt:      time.Unix(0, set.Referrer.JoinedAt),
-				NumberOfEpoch: set.Referrer.NumberOfEpoch,
+				PartyID:        types.PartyID(set.Referrer.PartyId),
+				JoinedAt:       time.Unix(0, set.Referrer.JoinedAt),
+				StartedAtEpoch: set.Referrer.StartedAtEpoch,
 			},
 		}
 
-		e.referrers[set.Referrer.PartyId] = set.Id
+		e.referrers[types.PartyID(set.Referrer.PartyId)] = setID
 
 		for _, r := range set.Referees {
-			e.referees[r.PartyId] = set.Id
+			partyID := types.PartyID(r.PartyId)
+			e.referees[partyID] = setID
 			newSet.Referees = append(newSet.Referees,
 				&types.Membership{
-					PartyID:       types.PartyID(r.PartyId),
-					JoinedAt:      time.Unix(0, r.JoinedAt),
-					NumberOfEpoch: r.NumberOfEpoch,
+					PartyID:        partyID,
+					JoinedAt:       time.Unix(0, r.JoinedAt),
+					StartedAtEpoch: r.StartedAtEpoch,
 				},
 			)
 		}
 
-		e.sets[set.Id] = newSet
+		e.sets[setID] = newSet
 	}
 }
 
-func NewEngine(epochEngine EpochEngine, broker Broker, timeSvc TimeService) *Engine {
+func (e *Engine) computeReferralSetsNotionalRunningVolume(epoch types.Epoch) {
```

Comment:
for this to work you need to make sure the referral engine is notified on end of epoch *after* market activity tracker

Reply:
Yes.

Reply:
That's already the case but I think linking an engine to the epoch service during its initialization reduces flexibility in terms of lifecycle. I will move it outside and add a comment, so we don't forget.